### DISCUSSION TITLE
Collegamento ad HA da remoto con HA, duckdns e caddy installati come container su docker

### DISCUSSION THREAD

**Discussion by jovannig** [indomus/forum#1507] [HUB: Home Assistant]
Ho seguito la guida per accedere da remoto ad HA, pero' la mia installazione e' diversa e non so come ottenere i certificati.
Uso Pi4 con docker e ho installato HA, duckdns e caddy come containers.
Ha e duckdns funzionano perfettamente, ma non riesco a capire come faccio ad ottenere i certificati (certfile e keyfile).
Potete aiutarmi, grazie!

**Comment by snakuzzo** [comment]
E' scritto tutto nella guida per Caddy....i certificati li richiede in automatico

**Comment by jovannig** [comment]
Ho riprovato adesso.
Quando creo il container per caddy eseguo i seguenti passi:
docker run --init -d --name="caddy" -e "TZ=Europe/Rome" -v /home/pi/caddy/Caddyfile:/etc/caddy/Caddyfile -v /home/pi/caddy/data:/data --restart always -p 443:443 -p 80:80 caddy:latest

docker logs caddy
Dopo questo comando ottengo degli errori:
`
{"level":"info","ts":1645005757.1144483,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":"caddyfile"}
{"level":"warn","ts":1645005757.1178417,"msg":"input is not formatted with 'caddy fmt'","adapter":"caddyfile","file":"/etc/caddy/Caddyfile","line":3}
{"level":"info","ts":1645005757.1201909,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["127.0.0.1:2019","localhost:2019","[::1]:2019"]}
{"level":"info","ts":1645005757.120696,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0x2d820a0"}
{"level":"info","ts":1645005757.1207657,"logger":"http","msg":"server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS","server_name":"srv0","https_port":443}
{"level":"info","ts":1645005757.12087,"logger":"http","msg":"enabling automatic HTTP->HTTPS redirects","server_name":"srv0"}
{"level":"info","ts":1645005757.122145,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["jovannig.duckdns.org"]}
{"level":"info","ts":1645005757.12278,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/data/caddy"}
{"level":"info","ts":1645005757.122809,"msg":"autosaved config (load with --resume flag)","file":"/config/caddy/autosave.json"}
{"level":"info","ts":1645005757.1228561,"logger":"tls","msg":"finished cleaning storage units"}
{"level":"info","ts":1645005757.1228437,"msg":"serving initial configuration"}
{"level":"info","ts":1645005757.1242177,"logger":"tls.obtain","msg":"acquiring lock","identifier":"jovannig.duckdns.org"}
{"level":"info","ts":1645005757.131414,"logger":"tls.obtain","msg":"lock acquired","identifier":"jovannig.duckdns.org"}
{"level":"info","ts":1645005760.423505,"logger":"tls.issuance.acme","msg":"waiting on internal rate limiter","identifiers":["jovannig.duckdns.org"],"ca":"https://acme-staging-v02.api.letsencrypt.org/directory","account":""}
{"level":"info","ts":1645005760.4236188,"logger":"tls.issuance.acme","msg":"done waiting on internal rate limiter","identifiers":["jovannig.duckdns.org"],"ca":"https://acme-staging-v02.api.letsencrypt.org/directory","account":""}
{"level":"info","ts":1645005761.622356,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"jovannig.duckdns.org","challenge_type":"http-01","ca":"https://acme-staging-v02.api.letsencrypt.org/directory"}
{"level":"error","ts":1645005772.339758,"logger":"tls.issuance.acme.acme_client","msg":"challenge failed","identifier":"jovannig.duckdns.org","challenge_type":"http-01","problem":{"type":"urn:ietf:params:acme:error:connection","title":"","detail":"Fetching http://jovannig.duckdns.org/.well-known/acme-challenge/9uUx-qKXL_g4-ZcgEcW8BAIQv4sOcMMFxCVKuuzn5-I: Timeout during connect (likely firewall problem)","instance":"","subproblems":[]}}
{"level":"error","ts":1645005772.3399515,"logger":"tls.issuance.acme.acme_client","msg":"validating authorization","identifier":"jovannig.duckdns.org","problem":{"type":"urn:ietf:params:acme:error:connection","title":"","detail":"Fetching http://jovannig.duckdns.org/.well-known/acme-challenge/9uUx-qKXL_g4-ZcgEcW8BAIQv4sOcMMFxCVKuuzn5-I: Timeout during connect (likely firewall problem)","instance":"","subproblems":[]},"order":"https://acme-staging-v02.api.letsencrypt.org/acme/order/44302448/1802399358","attempt":1,"max_attempts":3}
{"level":"info","ts":1645005773.7053802,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"jovannig.duckdns.org","challenge_type":"tls-alpn-01","ca":"https://acme-staging-v02.api.letsencrypt.org/directory"}
{"level":"error","ts":1645005784.4363174,"logger":"tls.issuance.acme.acme_client","msg":"challenge failed","identifier":"jovannig.duckdns.org","challenge_type":"tls-alpn-01","problem":{"type":"urn:ietf:params:acme:error:connection","title":"","detail":"Timeout during connect (likely firewall problem)","instance":"","subproblems":[]}}
{"level":"error","ts":1645005784.4367325,"logger":"tls.issuance.acme.acme_client","msg":"validating authorization","identifier":"jovannig.duckdns.org","problem":{"type":"urn:ietf:params:acme:error:connection","title":"","detail":"Timeout during connect (likely firewall problem)","instance":"","subproblems":[]},"order":"https://acme-staging-v02.api.letsencrypt.org/acme/order/44302448/1802400358","attempt":2,"max_attempts":3}
{"level":"error","ts":1645005785.9746382,"logger":"tls.obtain","msg":"could not get certificate from issuer","identifier":"jovannig.duckdns.org","issuer":"acme-staging-v02.api.letsencrypt.org-directory","error":"[jovannig.duckdns.org] solving challenges: jovannig.duckdns.org: no solvers available for remaining challenges (configured=[http-01 tls-alpn-01] offered=[http-01 dns-01 tls-alpn-01] remaining=[dns-01]) (order=https://acme-staging-v02.api.letsencrypt.org/acme/order/44302448/1802401708) (ca=https://acme-staging-v02.api.letsencrypt.org/directory)"}
{"level":"error","ts":1645005785.9747999,"logger":"tls.obtain","msg":"will retry","error":"[jovannig.duckdns.org] Obtain: [jovannig.duckdns.org] solving challenges: jovannig.duckdns.org: no solvers available for remaining challenges (configured=[http-01 tls-alpn-01] offered=[http-01 dns-01 tls-alpn-01] remaining=[dns-01]) (order=https://acme-staging-v02.api.letsencrypt.org/acme/order/44302448/1802401708) (ca=https://acme-staging-v02.api.letsencrypt.org/directory)","attempt":1,"retrying_in":60,"elapsed":28.843233734,"max_duration":2592000} 
`

**Reply by snakuzzo** [reply]
Ti prego...non creare n thread in questa discussione...che altrimenti non si capisce niente.
Comunque...innanzi tutto ti consiglio vivamente di passare all'utilizzo di docker-compose per la creazione dei container.
Quanto ai problemi...
1. Verifica che effettivamente il tuo dominio sia risolto con l'IP corretto
2. Sei sicuro di aver creato correttamente le regole di port forwarding ? 
Sul router devi creare almeno una regola per le challenge HTTP-01 (80 TCP -> 80 server su cui hai Caddy) o TLS-APLN-01 (443 TCP -> 443 server su cui hai Caddy).

**Comment by jovannig** [comment]
Se provo ad seguire il comando
ls -latr /home/pi/caddy/data/caddy/certificates/acme-staging-v02.api.letsencrypt.org-directory/jovannig.duckdns.org
mi da' errore:
ls: cannot access '/home/pi/caddy/data/caddy/certificates/acme-staging-v02.api.letsencrypt.org-directory/jovannig.duckdns.org': Permission denied

**Reply by indomus** [reply]
prova con `sudo ls`

**Comment by jovannig** [comment]
Ho provato, ma niente. 
Prima mi ha dato quegli errori che ho segnalato..

sudo ls -latr /home/pi/caddy/data/caddy/certificates/acme-staging-v02.api.letsencrypt.org-directory/jovannig.duckdns.org
ls: cannot access '/home/pi/caddy/data/caddy/certificates/acme-staging-v02.api.letsencrypt.org-directory/jovannig.duckdns.org': No such file or directory

**Comment by jovannig** [comment]
Ho cancellato immagini e container relativi a caddy.
Ho utilizzato docker-compose.yaml e ho anche cambiato nome sottodominio su duckdns.
Ma se digito docker logs caddy ottengo questo errore come riga finale:
{"level":"info","ts":1645018199.8838615,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"gagliardiphotography.duckdns.org","challenge_type":"http-01","ca":"https://acme-staging-v02.api.letsencrypt.org/directory"}

Sul router fritzbox ho fatto il port forwarding della 80, 443 e 8123.
